packages/labs/analyzer/src/lib/references.ts

```diff
@@ -66,90 +101,236 @@ export function getReferenceForSymbol(
+      const module = getModuleInfo(sourceFilePath, analyzer);
+      refPackage = module.packageJson.name;
+      refModule = path.join(path.dirname(module.jsPath), specifier);
+    } else if (specifier[0] === '/') {
+      // Absolute import; no package, just use the entire path as the
+      // module
+      refPackage = '';
+      refModule = specifier;
+    } else {
+      // External import: extract the npm package (taking care to respect
+      // npm orgs) and module specifier (if any)
+      const info = specifier.match(npmModule);
+      if (!info || !info.groups) {
+        throw new DiagnosticsError(
+          specifierExpression,
+          `External npm package could not be parsed from module specifier '${specifier}'.`
+        );
+      }
+      refPackage = info.groups.package;
+      refModule = info.groups.module || undefined;
+    }
+  }
+  return new Reference({
+    name,
+    package: refPackage,
+    module: refModule,
+    dereference: () =>
+      getExportFromSourcePath(
+        getPathForModuleSpecifierExpression(specifierExpression!, analyzer),
+        name,
+        analyzer
+      ),
+  });
+};
+
+/**
+ * Returns a `Reference` to a symbol declared in the current source file.
+ */
+const getLocalReference = (
+  location: ts.Node,
+  name: string,
+  analyzer: AnalyzerInterface
+) => {
+  const module = getModuleInfo(
+    location.getSourceFile().fileName as AbsolutePath,
+    analyzer
+  );
+  return new Reference({
+    name,
+    package: module.packageJson.name,
+    module: module.jsPath,
+    dereference: () =>
+      getExportFromSourcePath(
+        location.getSourceFile().fileName as AbsolutePath,
+        name,
+        analyzer
+      ),
+  });
+};
+
+/**
+ * For a given export declaration statement with named exports, returns an array
+ * of objects mapping the export name to a LocalNameOrReference, which is a
+ * string name that can be looked up directly in `getDeclaration()` of the
+ * declaring module for local declarations, or a `Reference` in the case of
+ * re-exported declarations.
+ *
+ * For example:
+ * ```
+ *   import {a} from 'foo';
+ *   const b = 'b';
+ *   const c = 'c';
+ *   export {a as x, b as y, c};
+ * ```
+ * This would return (using pseudo-code for Reference objects):
+ * ```
+ * [
+ *   {name: 'x', reference: new Reference('a', 'foo')},
+ *   {name: 'y', reference: 'b'},
+ *   {name: 'c', reference: 'c'},
+ * ]
+ * ```
+ * This also handles explicit re-export syntax, which all become References:
+ * ```
+ *   export {a as x, b as y, c} from 'foo';
+ * ```
+ */
+export const getExportReferences = (
+  exportDeclaration: ts.ExportDeclaration,
+  analyzer: AnalyzerInterface
+): Array<{exportName: string; decNameOrRef: LocalNameOrReference}> => {
+  const refs: Array<{exportName: string; decNameOrRef: string | Reference}> =
+    [];
+  const {exportClause, moduleSpecifier} = exportDeclaration;
+  // Note this function only handles named exports; wildcard exports must
```

Comment:
Should this be a TODO?

Reply:
No, the site that calls this function has a test that handles wildcard exports.

I refactored this to take the `exportClause` and `moduleSpecifier` directly (since they are deconstructed at the caller already), and required the `exportClause` to be defined in this function, making the need to check it moot since it's already checked at the callsite.